Add recursive object passing (includes arrays)

typical000 opened the pull request:
Added recursive object passing. Why:
1. Remove obligatory setting jss-nested plugin **BEFORE** jss-default-unit
2. Add possibility to set values with default units inside arrays or objects in terms of contrib pulgin jss-shorthand:
```javascript
import jss from 'jss'
import jssShorthand from 'jss-shorthand'

jss.use(jssShorthand())

let sheet = jss.createStyleSheet({
  container: {
    padding: [20, 10],
    background: {
        color: '#000',
        image: 'url(image.jpg)',
        position: [20, 20],
    },
  }
})
console.log(sheet.toString())
```
And output will be:
```css
.jss-0-0 {
  padding: 20px 10px;
  background: #000 url(image.jpg) 20px 20px;
}
```

kof commented on src/index.js:
```diff
@@ -1,19 +1,58 @@
 /**
+ * Recursive deep style passing function
+ *
+ * @param {(Object|Number|String)} value
```
didn't update?

kof commented on src/index.js:
```diff
@@ -1,19 +1,58 @@
+        passStyle(style[prop], options)
+        break
+      case Array:
+        for (const item in style[prop]) {
```
arrays should be iterated with for, right?

typical000 replied:
I didn't found any other solution considering *eslint*. For example `Array.prototype.forEach` or `Array.prototype.map` throws eslint error `no-loop-func`

kof replied:
I don't get it, maybe linter is broken.

typical000 replied:
Fixed. Check my last commit to pull request. Changed 'for' loop to this:
> style[prop] = style[prop].map(value => addUnit(prop, value, options))

kof commented on src/index.js:
```diff
@@ -1,19 +1,58 @@
+ * @param {Object} options
+ * @return {String} string with units
+ */
+function checkForNumber(prop, item, options) {
```
second argument should be called `value`

kof replied:
I would name this function `addUnit`

kof commented on src/index.js:
```diff
@@ -1,19 +1,58 @@
+export default function defaultUnit(options = { unit: 'px' }) {
   return rule => {
-    const {style, type} = rule
+    const { style, type } = rule
```
we don't use spaces around destructured variables

typical000 replied:
But if I remove spaces, linter throws errors like: 
> 52:48  error  A space is required after '{'   object-curly-spacing
> 52:58  error  A space is required after '}'   object-curly-spacing
It's ok?

kof commented on src/index.js:
```diff
@@ -1,19 +1,58 @@
  * @api public
  */
-export default function defaultUnit(options = {unit: 'px'}) {
+export default function defaultUnit(options = { unit: 'px' }) {
```
we don't use spaces around destructured variables

typical000 replied:
But if I remove spaces, linter throws errors like. It's ok?
> 52:48  error  A space is required after '{'   object-curly-spacing
> 52:58  error  A space is required after '}'   object-curly-spacing

kof commented on src/index.js:
```diff
@@ -1,19 +1,58 @@
+ * @param {Object} options
+ * @return {Object} resulting style
+ */
+function passStyle(style, options) {
```
I would name this function `iterate`

kof commented:
- [ ] We need to add tests for all this cases, can you do that?

kof commented:
I think we have got an unintentional linter update.

kof commented:
hmm https://github.com/jsstyles/eslint-config-jss/blob/master/eslintrc.json#L21 need to dig in why this option doesn't work

typical000 commented:
I just updated all dependencies in jss-default-unit to lastest versions:
```````  
"devDependencies": {
    "babel-cli": "^6.10.1",
    "babel-core": "^6.10.4",
    "babel-eslint": "^6.1.0",
    "babel-loader": "^6.2.4",
    "babel-preset-es2015": "^6.9.0",
    "babel-preset-stage-0": "^6.5.0",
    "eslint": "^2.13.1",
    "eslint-config-airbnb": "^8.0.0",
    "eslint-config-jss": "^1.1.0",
    "eslint-plugin-import": "^1.10.2",
    "eslint-plugin-jsx-a11y": "^1.5.5",
    "eslint-plugin-react": "^5.2.2",
    "jss": "^4.0.3",
    "opener": "^1.4.1",
    "qunitjs": "^2.0.0",
    "rimraf": "^2.5.3",
    "webpack": "^1.13.1"
  },
`````````````
And linter works fine, except errors about quoted properties in **numericProps.js**:

>   11:3  error  Unnecessarily quoted property 'flex' found     quote-props
  22:3  error  Unnecessarily quoted property 'opacity' found  quote-props
  23:3  error  Unnecessarily quoted property 'order' found    quote-props
  24:3  error  Unnecessarily quoted property 'orphans' found  quote-props
  26:3  error  Unnecessarily quoted property 'widows' found   quote-props
  28:3  error  Unnecessarily quoted property 'zoom' found     quote-props

I can update dependencies in terms of my pull request. But I neet to update code from tests too. Because **qunit** and **jss** are updated too and we recieve errors like:

> 
Expected: 	
"@media print {
  .button--jss-0-4 {
    border-left: 1px;
    border: 3px;
  }
}"
Result: 	
"@media print {
  .button-590492203 {
    border-left: 1px;
    border: 3px;
  }
}"
Diff: 	
"@media print {
  .button--jss-0-4590492203 {
    border-left: 1px;
    border: 3px;
  }
}"

typical000 commented:
> We need to add tests for all this cases, can you do that?
Yes. I'll try to add tests. But some things we can't test - like deep passing 'objects that contains arrays of objects' without my **jss-expand** plugin (without it we recieve only 'undefined')

kof commented on src/index.js:
```diff
@@ -1,6 +1,47 @@
+        iterate(style[prop], options)
+        break
+      case Array:
+        style[prop] = style[prop].map(value => addUnit(prop, value, options))
```
I have been using for loops because they are still faster and this is a kinda hot code.

typical000 replied:
So, replace back to 'for-in' loop?

kof replied:
Why for-in? Why do you want to iterate an array with it? Why not
```
        for (let i = 0; i < style[prop].length; i++) {
          style[prop][i] = addUnit(prop, style[prop][i], options)
        }
```

typical000 replied:
True, why not. I'll change it

typical000 commented:
@kof About tests: the only one test that we can added for now (without 'fallbacks' feature released) is something like this:
``````````````
QUnit.test('values in arrays', function (assert) {
  var ss = jss.createStyleSheet({
    a: {padding: [10, 15]}
  }, {named: false})
  assert.equal(ss.toString(), 'a {\n  padding: 10px;\n  padding: 15px;\n}', 'px added')
})
``````````````

Objects passing, without **jss-expand** returns `something: [object Object]`. And this is not a good thing that we can add as tests :)

kof commented:
Ok then we need to release it first.

kof commented:
Released version 5, lets add tests.

typical000 commented:
1. Updated dependencies and, consequently, the code
2. Added test for numbers in arrays, and for numbers inside objects

kof commented on src/index.js:
```diff
@@ -1,6 +1,49 @@
+ *
+ * @param {Object} original style
+ * @param {Object} options
+ * @return {Object} resulting style
```
actually we don't need to return, because we mutate, return makes it look like immutable, but it is not.

typical000 replied:
Removed in https://github.com/jsstyles/jss-default-unit/pull/6/commits/0257e38dda1ef0a1456de6e05711b12c329f3393

kof commented:
We currently compile `prop: [1, 2]` to `prop: '1px,2px'` which is a multi value feature of css, for e.g. allows to specify multiple backgrounds.

However the only way to specify  space separated shortcut is currently a string: `margin: '10px 10px'`, it would be nice to allow default unit also for shortcut notation. Any ideas?

One I can think of is: `margin: [[10, 10]]` which is essentially explicit multi value syntax where we can treat the first value as a space separated shortcut.

kof commented:
The other possibility is to leave this out and let only expanded syntax have default-unit enabled, so if you want to get `margin: 10px 5px` you would need to write `margin: {top: 10, right: 5, bottom: 10, left: 5}`.

kof commented:
`margin: [[10, 10]]` is not very nice, but still shorter than `margin: {top: 10, right: 5, bottom: 10, left: 5}`

typical000 commented:
My plugin solves situation when we need to write space-separated values https://github.com/typical000/jss-expand there are config that indicates to print 'comma-separated' or 'space-separated' values. Part of config:
````````````````````js
/*
 * Scheme for converting properties from array to nomral style.
 * TRUE - indicates that styles will be joined with whitespace. E.g:
 *        "padding: [ 10, 20 ]" => "padding: 10 20;"
 * FALSE - indeicates that styles will be joined with comma. E.g:
 *         "box-shadow: [ '0 0 5px red', '0 0 2px green' ]" => "box-shadow: 0 0 5px red, 0 0 2px green"
 */
export const propArray = {
  'margin': true,
  'padding': true,
  'border-radius': true,
  'background-size': true,
  'background-position': true,
  'background': false,
  'transform': false,
  'transition': false,
  'animation': false,
  'box-shadow': false,
  'text-shadow': false,
}
````````````````````
So, using **jss-expand** you can write
````````````````````js
margin: [10, 5]
````````````````````
OR
````````````````````js
margin: {top: 10, right: 5, bottom: 10, left: 5}
````````````````````

kof commented:
I think such a config is a bad idea, because it will have lots of broken edge cases and will need to be maintained all the time. I am not even aware of how much props have the shortcut syntax and whether this list is growing or not and how to keep it updated.

kof commented:
Just one example: `border-right`, what would jss-expand do in this case?

kof commented:
Another problem with jss-expand: `margin: ['5px', '10px']` without jss-expand compiles to `margin: 5px,10px`, same with jss-expand would compile to `margin: 5px 10px`. I think this will lead to a lot of confusion.

typical000 commented:
> I think such a config is a bad idea
I agree with you. But I haven't find any 'general rule' how to distinguish values, that must be printed 'comma-separated' and 'space-separated'. 
And the main goal was - 'one syntax for all styles'. So, I haven't find any other solution than config.

> Just one example: border-right, what would jss-expand do in this case?

````````````````````````js
a: {
  'border-right': {
    width: '5px',
    style: 'solid',
    color: '#f00'
  }
}
````````````````````````
result:
````````````````````````css
a {
  border-right: 5px solid #f00;
}
````````````````````````

> Another problem with jss-expand: margin: ['5px', '10px'] without jss-expand compiles to margin: 5px,10px, same with jss-expand installed would compile to margin: 5px 10px. I think this will lead to a lot of confusion.

But writing `margin: ['5px', '10px']` without **jss-expand** makes absolutely no sense.

kof commented:
> But writing margin: ['5px', '10px'] without jss-expand makes absolutely no sense.

It is an undocumented value, just by accident because `[1,2].toString()` results in '1,2' we got the multi value behavior.

kof commented:
Also, this result with "border-right", have you tried it? Because I don't see border-right in the list in props.js, I didn't fully read the code though.

Another question: `box-shadow: [0, 0, 5, 'red']` would result in `box-shadow: 0,0,5,red`, no?

typical000 commented:
Nope. There are 2 main configs (**propArray** and **propObject**), the last one **porpArrayInObj** is a subsidiary config only for arrays used inside objects. File is here: https://github.com/typical000/jss-expand/blob/master/src/props.js

In brief:
* **propArray** handles listed inside keys with values, that are arrays
``````````````````````````js
export const propArray = {
  'margin': true,
  'padding': true,
  ...
  'box-shadow': false,
  'text-shadow': false,
}
``````````````````````````
if *TRUE* - array will be join by space. If *FALSE* - with comma. So you can write:
``````````````````````````js
a: {
      padding: [ 20, 10 ],
      transition: [ 'opacity 1s linear', 'transform 300ms ease' ]
    }
``````````````````````````
resulting css:
``````````````````````````css
a {
    padding: 20 10;
    transition: opacity 1s linear, transform 300ms ease;
}
``````````````````````````


* **propObj** handles styles that are written on *object* form. Part of config
``````````````````````````js
export const propObj = {
  padding: {
    template: 'top right bottom left',
    prop: {
      top: 0,
      right: 0,
      bottom: 0,
      left: 0,
    }
  },
  ...
  'box-shadow': {
    template: 'x y blur spread color inset',
    prop: {
      x: 0,
      y: 0,
      blur: '',
      spread: '',
      color: '',
      inset: '',
    }
  },
}  
``````````````````````````
Here:
1. *template* - how to build value. Because order does matter
2. *prop* - set of accepted values, by W3C
And example:
``````````````````````````js
    a: {
      border: {
        width: '1px',
        style: 'solid',
        color: '#f00'
      },
      boxShadow: {
        x: 0,
        y: 0,
        blur: '20px',
        color: 'red'
      }
   },
``````````````````````````
And resulting css:
``````````````````````````css
a {
    border: 1px solid #f00;
    box-shadow: 0 0 20px red;
}
``````````````````````````

So in terms of your question: `box-shadow: [0, 0, 5, 'red']` would result  `box-shadow: 0,0,5,red` - yes. In case of **jss-expand** you must write:
``````````````````````````js
// For single shadow:
boxShadow: {
  x: 0,
  y: 0,
  blur: '20px',
  color: 'red'
}
// OR you can write as before
boxShadow: '0 0 20px red';


// For multi box-shadow you can write:
boxShadow: [{
  x: 0,
  y: 0,
  blur: '20px',
  color: 'red'
}, {
  x: 0,
  y: 0,
  blur: '10px',
  color: 'blue'
}]
// OR as more simple array
boxShadow: [
  '0 0 20px red',
  '0 0 10px blue'
]
``````````````````````````

kof commented:
ok still the question if it makes sense to support shorthand syntax with numbers for both comma and space separated values.

kof commented:
Does this make sense to you: `margin: [[10, 5]]` ?

typical000 commented:
> Does this make sense to you: margin: [[10, 5]] ?

From one point of view if makes sense, because:
* There is no need for additional plugin for this simple feature, if it became part of jss core
* If it will be documented with 'comma-separated' syntax like `margin: [10, 5]`

And contra:

* It looks like 'the only possibility' to implement 'space-separated' syntax considering that 'comma-separated' syntax is written with simple arrays
* In **jss-expand** the main goal was **one syntax for all**. But there are extra evaluations. And as you say - "I think such a config is a bad idea" and I agree with you.

But, I think that is **matter of habit**. If you implement this syntax and document it - there are no problems to use it if you know what styles must be written in 'comma-separated' or 'space-separated' form. 

And, If you implement it - I can rewrite **jss-expand** and remove arrays configs, leaving only expanded object syntax, and plugin will start exactly correspont to his name - **expand** :)

kof commented:
The only problem with jss-expand so far: properties which can have comma separated values like a background, can't be written in a short way with numbers for space separated value.

kof commented:
And to fix that, I need to give `margin: [[5, 10]]` a go in in the core.

kof commented:
Lets do that.

kof commented:
Released https://github.com/jsstyles/jss/issues/264

kof commented:
We need to change:

- jss version
- support array inside of array like [this](https://github.com/jsstyles/jss/commit/267f849f372edd0d8cb0c2107fe6bc92e0cc93ba#diff-2b4ca49d4bb0a774c4d4c1672d7aa781R61)

kof commented:
I have already added documentation, we need to release this one quickly https://github.com/jsstyles/jss/blob/master/docs/json-api.md#alternative-syntax-for-space-and-comma-separated-values

typical000 commented:
@kof 
* updated *numbericProps.js* for avoiding 'quote-props' errors
* updated dependencies up to 5.1.0
* updated code
* added tests for comma-separated syntax and space-separated syntax

kof commented:
released
